master/internal/provisioner/scale_decider.go

```diff
@@ -110,15 +124,47 @@ func (s *scaleDecider) findInstancesToTerminate(
 			toTerminate[inst.ID] = true
 
 		case Running:
-			if _, ok := s.agentSnapshot[inst.AgentName]; ok {
+			if _, ok := s.idleAgentSnapshot[inst.AgentName]; ok {
 				idleInstances[inst.ID] = true
 			}
 		}
 	}
 
+	// Identify instances that are not currently connected to the master.
+	now := time.Now()
+	for _, inst := range s.instanceSnapshot {
+		// If instance is connected no need to do anything here.
+		if _, connected := s.connectedAgentSnapshot[inst.AgentName]; connected {
+			continue
+		}
+
+		// If instance instance is still in the start-up period, do not terminate it for
+		// being disconnected.
+		if inst.LaunchTime.Add(s.maxStartingPeriod).After(now) {
+			continue
+		}
+
+		// Don't terminate instances that are already stopped or are stopping.
+		switch inst.State {
+		case Stopping, Stopped:
+			continue
+		}
+
+		disconnectedInstances[inst.ID] = true
+	}
+
+	// Terminate instances that have not connected to the master for a long time.
+	var longUnconnected map[string]bool
+	s.pastDisconnectedInstances, longUnconnected = findInstancesLongInSameState(
+		s.pastDisconnectedInstances, disconnectedInstances, s.maxIdlePeriod)
```

Comment:
blocking: I think we should add one more field called `maxDisconnectedPeriod` in the configuration rather than reusing `maxIdlePeriod or at least use a constant.

Reply:
made it a const, don't think this is something users need / should configure